TestInvalidEntryPoint failing following release of importlib_metadata 8.7

In https://github.com/python/importlib_metadata/pull/518, `importlib_metadata` started adding some validation for invalid entry points. Now they're no longer allowed to be constructed and will raise a ValueError if invalid input is presented.

This new behavior clashes with Setuptools' own expectation that the entry points will fail lazily when relevant attributes are accessed and what exception is thrown.

This change seems to have blocked the Setuptools 80 release because somehow `importlib_metadata` is getting used at head, presumably because some other package is pulling it in.

```
__________________ TestWriteEntries.test_invalid_entry_point ___________________
[gw0] darwin -- Python 3.9.13 /Users/runner/work/setuptools/setuptools/.tox/py/bin/python

self = <setuptools.tests.test_egg_info.TestWriteEntries object at 0x106123160>
tmpdir_cwd = local('/Users/runner/work/setuptools/setuptools')
env = '/var/folders/7g/cnfrzbd555q9lmpsr1byngs80000gn/T/setuptools-test.a8lus_rf'

    def test_invalid_entry_point(self, tmpdir_cwd, env):
        dist = Distribution({"name": "foo", "version": "0.0.1"})
        dist.entry_points = {"foo": "foo = invalid-identifier:foo"}
        cmd = dist.get_command_obj("egg_info")
        expected_msg = r"Problems to parse .*invalid-identifier.*"
        with pytest.raises(errors.OptionError, match=expected_msg) as ex:
>           write_entries(cmd, "entry_points", "entry_points.txt")

/Users/runner/work/setuptools/setuptools/setuptools/tests/test_egg_info.py:1292: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Users/runner/work/setuptools/setuptools/setuptools/command/egg_info.py:704: in write_entries
    eps = _entry_points.load(cmd.distribution.entry_points)
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/functools.py:888: in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
/Users/runner/work/setuptools/setuptools/setuptools/_entry_points.py:61: in load
    return validate(metadata.EntryPoints(groups))
/Users/runner/work/setuptools/setuptools/.tox/py/lib/python3.9/site-packages/importlib_metadata/__init__.py:366: in <genexpr>
    EntryPoint(name=item.value.name, value=item.value.value, group=item.name)
/Users/runner/work/setuptools/setuptools/.tox/py/lib/python3.9/site-packages/importlib_metadata/__init__.py:214: in __init__
    self.module
/Users/runner/work/setuptools/setuptools/.tox/py/lib/python3.9/site-packages/importlib_metadata/__init__.py:227: in module
    return self._match.module
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/functools.py:993: in __get__
    val = self.func(instance)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = EntryPoint(name='foo', value='invalid-identifier:foo', group='foo')

    @functools.cached_property
    def _match(self) -> _EntryPointMatch:
        match = self.pattern.match(self.value)
        if not match:
>           raise ValueError(
                'Invalid object reference. '
                'See https://packaging.python.org/'
                '/en/latest/specifications/entry-points/#data-model',
                self.value,
            )
E           ValueError: ('Invalid object reference. See https://packaging.python.org/en/latest/specifications/entry-points/#data-model', 'invalid-identifier:foo')

/Users/runner/work/setuptools/setuptools/.tox/py/lib/python3.9/site-packages/importlib_metadata/__init__.py:241: ValueError
```

--- a/newsfragments/4961.bugfix.rst
+++ b/newsfragments/4961.bugfix.rst
@@ -1,0 +1,1 @@
+Update test to honor new behavior in importlib_metadata 8.7.

--- a/setuptools/_entry_points.py
+++ b/setuptools/_entry_points.py
@@ -15,6 +15,10 @@ def ensure_valid(ep):
     """
     Exercise one of the dynamic properties to trigger
     the pattern match.
+
+    This function is deprecated in favor of importlib_metadata 8.7 and
+    Python 3.14 importlib.metadata, which validates entry points on
+    construction.
     """
     try:
         ep.extras

--- a/setuptools/tests/test_egg_info.py
+++ b/setuptools/tests/test_egg_info.py
@@ -1287,10 +1287,11 @@ class TestWriteEntries:
         dist = Distribution({"name": "foo", "version": "0.0.1"})
         dist.entry_points = {"foo": "foo = invalid-identifier:foo"}
         cmd = dist.get_command_obj("egg_info")
-        expected_msg = r"Problems to parse .*invalid-identifier.*"
-        with pytest.raises(errors.OptionError, match=expected_msg) as ex:
+        expected_msg = r"(Invalid object reference|Problems to parse)"
+        with pytest.raises((errors.OptionError, ValueError), match=expected_msg) as ex:
             write_entries(cmd, "entry_points", "entry_points.txt")
             assert "ensure entry-point follows the spec" in ex.value.args[0]
+            assert "invalid-identifier" in str(ex.value)
 
     def test_valid_entry_point(self, tmpdir_cwd, env):
         dist = Distribution({"name": "foo", "version": "0.0.1"})

